src/librustc_mir/const_eval/query.rs

```diff
@@ -0,0 +1,277 @@
+            // an error must be reported.
+            let v = err.report_as_error(ecx.tcx, "could not evaluate static initializer");
+            tcx.sess.delay_span_bug(
+                err.span,
+                &format!("static eval failure did not emit an error: {:#?}", v)
+            );
+            v
+        } else if def_id.is_local() {
+            // constant defined in this crate, we can figure out a lint level!
+            match tcx.def_kind(def_id) {
+                // constants never produce a hard error at the definition site. Anything else is
+                // a backwards compatibility hazard (and will break old versions of winapi for sure)
+                //
+                // note that validation may still cause a hard error on this very same constant,
+                // because any code that existed before validation could not have failed validation
+                // thus preventing such a hard error from being a backwards compatibility hazard
+                Some(DefKind::Const) | Some(DefKind::AssocConst) => {
+                    let hir_id = tcx.hir().as_local_hir_id(def_id).unwrap();
+                    err.report_as_lint(
+                        tcx.at(tcx.def_span(def_id)),
+                        "any use of this value will cause an error",
+                        hir_id,
+                        Some(err.span),
+                    )
+                },
+                // promoting runtime code is only allowed to error if it references broken constants
+                // any other kind of error will be reported to the user as a deny-by-default lint
+                _ => if let Some(p) = cid.promoted {
+                    let span = tcx.promoted_mir(def_id)[p].span;
+                    if let err_inval!(ReferencedConstant) = err.error {
+                        err.report_as_error(
+                            tcx.at(span),
+                            "evaluation of constant expression failed",
+                        )
+                    } else {
+                        err.report_as_lint(
+                            tcx.at(span),
+                            "reaching this expression at runtime will panic or abort",
+                            tcx.hir().as_local_hir_id(def_id).unwrap(),
+                            Some(err.span),
+                        )
+                    }
+                // anything else (array lengths, enum initializers, constant patterns) are reported
+                // as hard errors
+                } else {
+                    err.report_as_error(
+                        ecx.tcx,
+                        "evaluation of constant value failed",
+                    )
+                },
+            }
+        } else {
+            // use of broken constant from other crate
+            err.report_as_error(ecx.tcx, "could not evaluate constant")
+        }
+    })
+}
+
+// Returns a pointer to where the result lives
+fn eval_body_using_ecx<'mir, 'tcx>(
```

Comment:
Please move this up ("definition before use").

Reply:
:( This is rust where we don't have the ordering problem and I can just Ctrl + Click in an IDE and it will jump me to the right thing. Is there really some benefit to manually ordering things?

Reply:
Which IDE actually makes that work for rustc source code? I haven't found one. Even vscode on Miri (let alone rustc itself) doesn't make that work most of the time as the toolchain is self-compiled and hence comes without RLS.  I do all my rustc hacking in a plain editor (Kate) and also have to read code in GH regularly, and Ctrl-Click doesn't work in either of them.

I mean you can also pick random variable names and code can still compile. Just because the compiler lets you do something doesn't mean that code readability can't be improved. When readining a file, it IMO helps quite a lot if functions are not in an entirely random order.

Reply:
rust analyzer + vscode works mostly. Some things it won't grok, but it's been a superb help so far.

> Also I have to read code in GH regularly. Ctrl-Click doesn't work in either of them.

Ok that's a pretty good reason.

> When readining a file, it IMO helps quite a lot if functions are not in an entirely random order.

When you put it like that... I now feel pretty stupid for my snarky "argument" in my previous post.

Reply:
> rust analyzer + vscode works mostly. Some things it won't grok, but it's been a superb help so far.

RA + VSCode is pretty fantastic when working on rustc imo, at least on non-macro-heavy code. I highly recommend it.